File: 프로그래머스/레벨 2/프로그래머스  가장 큰 정사각형 찾기.cpp


#include <iostream>
#include<vector>
#include <algorithm>

using namespace std;

int solution(vector<vector<int>> board)
{
    int answer = board[0][0];
    
    for(int i=1;i<board.size();i++){
        for(int j=1;j<board[0].size();j++){
            if(board[i][j] == 1){
                board[i][j] = 1 + min({board[i-1][j-1], board[i][j-1], board[i-1][j]});
                answer = max(answer, board[i][j]);
            }
        }
    }
    
    

    return answer*answer;
}

// 처음에는 내가 첫 점, 오른쪽 점과 밑의 점이 1, 대각선 밑도 1일 경우에 그 대각선 밑의 수를
// 증가시키려고 했는데, seg 에러가 떠서

// 다음 방법은 min을 활용하여 +1,0 0,+1 +1,+1을 이용하는 것

// 좀 더 많이 풀어보자

File: 프로그래머스/레벨 2/프로그래머스 가장 큰 정사각형 찾기.py
def solution(board):
    answer = 1234
    
    n = len(board)
    m = len(board[0])
    # dp = [[1] * (m+1) for _ in range(n+1)]
    
    Max = board[0][0]
    for i in range(1,n):
        for j in range(1,m):
            if board[i][j] == 1:
                board[i][j] = 1 + min(board[i][j-1], board[i-1][j], board[i-1][j-1])
                Max = max(Max, board[i][j])

    return Max * Max
